node/src/common/JXCoreEngine.cpp

```diff
@@ -0,0 +1,456 @@
+}
+
+/// Callback invoked when JavaScript code calls a function that was registered as a call from script.
+void JXCallCallback(JXValue* argv, int argc)
+{
+    if (argc != 2)
+    {
+        LogWarning("Invalid call callback.");
+        return;
+    }
+
+    const char* callIdHex = JX_GetString(argv);
+    unsigned long long callId = std::strtoull(callIdHex, nullptr, 16);
+    if (callId == 0)
+    {
+        LogWarning("Invalid result callback ID.");
+        return;
+    }
+
+    const char* argsJson = JX_GetString(argv + 1);
+
+    std::function<void(std::string)>* callbackPtr = reinterpret_cast<std::function<void(std::string)>*>(callId);
+    try
+    {
+        (*callbackPtr)(argsJson ? std::string(argsJson) : std::string());
+    }
+    catch (...)
+    {
+        LogWarning("Script call callback function threw an exception.");
+    }
+
+    // Don't delete this callback function; it may be invoked multiple times.
+}
+
+inline void LogErrorAndThrow(const char* message)
+{
+    LogError(message);
+    throw std::logic_error(message);
+}
+
+// Static member initialization
+std::once_flag JXCoreEngine::_initOnce;
+std::string JXCoreEngine::_workingDirectory;
+
+JXCoreEngine::JXCoreEngine() :
+    _started(false),
+    _callScriptFunction(nullptr)
+{
+    _dispatcher.Initialize();
+}
+
+JXCoreEngine::~JXCoreEngine()
+{
+    _dispatcher.Shutdown();
+}
+
+void JXCoreEngine::DefineScriptFile(std::string scriptFileName, std::string scriptCode)
+{
+    LogTrace("JXCoreEngine::DefineScriptFile(\"%s\", \"...\")", scriptFileName.c_str());
+
+    if (scriptFileName == mainScriptFileName)
+    {
+        throw new std::invalid_argument("Invalid script file name: 'main.js' is a reserved name.");
+    }
+
+    _dispatcher.Dispatch([this, scriptFileName, scriptCode]()
+    {
+        if (!_started)
+        {
+            _initialScriptMap.emplace(scriptFileName, scriptCode);
+        }
+        else
+        {
+            JX_DefineFile(scriptFileName.c_str(), scriptCode.c_str());
+        }
+    });
+}
+
+void JXCoreEngine::Start(std::string workingDirectory, std::function<void(std::exception_ptr ex)> callback)
+{
+    LogTrace("JXCoreEngine::Start(\"%s\")", workingDirectory.c_str());
+
+    if (workingDirectory.length() == 0)
+    {
+        throw std::invalid_argument("A working directory is required.");
+    }
+
+    if (_workingDirectory.length() == 0)
+    {
+        _workingDirectory = workingDirectory;
+    }
+    else if (workingDirectory != _workingDirectory)
+    {
+        // This limitation of JXCore is not represented in the INodeEngine interface (e.g. as a static
+        // initialization method taking the working directory), because other node engines might not
+        // have the same limitation.
+        LogErrorAndThrow("Cannot start multiple JXCore instances with different working directories.");
+    }
+
+    try
+    {
+        std::call_once(_initOnce, [=]()
+        {
+            JX_InitializeOnce(workingDirectory.c_str());
```

Comment:
FYI; On the new api there is no need for `JX_InitializeOnce` etc. JS_StartEngine is sufficient